flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -151,6 +149,9 @@
 	 * It must be single-threaded. Eventually it will be replaced by main thread executor. */
 	private final ScheduledExecutor timer;
 
+	@Nullable
+	private ComponentMainThreadExecutor mainThreadExecutor;
```

Comment:
It's not nice that this is not a `final` field, but I guess this would need large scope of changes, as there is the same issue in `ExecutionGraph`.

However, why does it have to be `@Nullable`? `ExecutionGraph#jobMasterMainThreadExecutor` uses interesting trick with `DummyComponentMainThreadExecutor`. I'm not sure if doing `instanceof` checks is better or not, but maybe for the sake of consistency we should do the same here?

Reply:
Yes, it could not be a `final` variable based on current implementation of `ExecutionGraph`. I'm just not a fan of the trick of `DummyComponentMainThreadExecutor`. But I agree with you that it's not a bad idea to provide a consistent way with `ExecutionGraph`. Let me think of it a bit more. Maybe we could find a better solution.

Reply:
Sadly I didn't find an easy way. It involves too much for this. For now I use the `DummyComponentMainThreadExecutor` and mark it `@Nonnull` as you advised. But I think it's worth a separate ticket. I put it on my follow-up list.